### ISSUE TITLE
Create an ".async" modifier for event handlers 

### ISSUE THREAD

**Issue by twitwi** [vuejs/vue#10338]
### What problem does this feature solve?
In an event handler, one can easily mix property modifications and method calls (which is great), as in:

    <span @click=" accumulator += fetchNextItem() ">do it</span>

However, if the called method is async, one need to actually wrap this code in an additional method like:

    <span @click=" fetchAndAccumulateNextItem() ">do it </span>
    ....
    methods: {
        async fetchAndAccumulateNextItem() {
            this.accumulator += await this.fetchNextItem()
        },
        async fetchNextItem() { .... } /* unmodified */
    }

This can be slightly inconvenient if there are many asynchronous methods of which we use the return value.
I think this might become more and more common as people start understanding and using async/await more and more.

Initial context: I have a very specific use case of a vuejs<->python bridge that makes (among other things) all the python-written method callable from vuejs, but as the call goes through websockets, all methods end up async. https://github.com/twitwi/vuejs-python

### What does the proposed API look like?
I'd suggest a ".async" modifier that would allow for "await" in the event handler snippet.
For the example above, it would be written as:

    <span @click.async=" accumulator += await fetchNextItem() ">do it</span>

(this is probably useful for all types of events, not only click)

**Comment by posva** [comment]
I don't think it's worth adding another modifier that only serves a very few uses cases: there are not many cases where the code is still readable when doing an `await` inside an event handler **in the html**. You don't need to create a method, you should be able to do:

```vue
<span @click=" async () => accumulator += await fetchNextItem() ">do it</span>
```

You could also create your own function helpers that are accessible globally like `asyncAdd`:

```vue
<span @click=" asyncAdd(accumulator, fetchNextItem())">do it</span>
```

which ends up being shorter than the proposal.

Let's wait for more feedback on this

**Comment by twitwi** [comment]
Thanks for the reply.
I like the first workaround, I'd give it a try

**Comment by twitwi** [comment]
:+1:
For some reason, I needed to add some parentheses around it, but it works well.
For the record, I give the result here.

For the example above, this yields:
    <span @click="(async () => accumulator += await fetchNextItem())">do it</span>

For my use case (with affecting an attribute), this gives:

    <a @click="(async()=> title = await fetchTitle())">fetch</a>

Or, alternatively, inspired by the second suggestion:

    <a @click="asyncSet('title', fetchTitle())">fetch</a>
    ....
    methods: {
        async asyncSet (attr, asyncVal) {
            this.$set(this, attr, await asyncVal)
        },
    }

Thanks again for the workarounds.

**Comment by sirlancelot** [comment]
@posva it looks like the parser doesn't understand async function expressions. Here's the compiler output:

```js
function render() {
  with(this) {
    return _c('span', {
      on: {
        "click": function ($event) {
          async () => accumulator += await fetchNextItem()
        }
      }
    }, [_v("do it")])
  }
}
```

[async inline function epxression demo](https://template-explorer.vuejs.org/#%3Cspan%20%40click%3D%22%20async%20()%20%3D%3E%20accumulator%20%2B%3D%20await%20fetchNextItem()%20%22%3Edo%20it%3C%2Fspan%3E)

It appears to nest the entire string inside a normal function. Should this issue be re-purposed or create a new one?

**Comment by posva** [comment]
@sirlancelot you are right, I adapted the compiler to allow it: https://github.com/vuejs/vue/pull/10361

**Comment by adjenks** [comment]
This works for me:
`@input="(async function(){myVariable= await _queryVals($event)})"`
but this failed:
`@input="(async ()=>{myVariable= await _queryVals($event)})"`

>Syntax Error: SyntaxError: await is a reserved word
